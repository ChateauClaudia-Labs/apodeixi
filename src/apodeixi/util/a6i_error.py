import sys                                              as _sys
import traceback                                        as _traceback
from io                                                 import StringIO

class FunctionalTrace():
    '''
    Data structure used to record useful information about errors. This can be thought as the functional analogue of
    a "stack trace". A stack trace shows the technical sequence of calls that leads to an error. This is not always useful
    to functionally identify what went wrong because it misses the functional context of what the code was supposed to
    be doing when the error was raised.

    To remedy this, the pattern used in Apodeixi is to use the `FunctionalTrace` object to record that functional context.
    In functional terms, the system is treated as if it was some sort of BPMN flow, with various functionally meaningful
    activities that connect sequentially, happen in parallel, fork into loops of lower-level activities, or join up.

    The `FunctionalTrace` data structure identifies the fragment of such BPMN logical flow that leads to the error.
    The BPNM could be thought as some kind of sorted tree. For a given node, its immediate children can be any of:
    
    * sub-activities ordered in the sequence in which they should happen.
    * a sub-activity that is looped over multiple times

    The `FunctionalTrace` records the branch of such conceptual BPNM tree, thus identifying top-down the functional
    intention that lead to wanting to execute the code for the leaf of such branch.

    @param path_mask A function that takes a string argument and returns a string. Normally it is None, but
                it is used in situations (such as in regression testing) when observability should not
                report the paths "as is", but with a mask. For example, this can be used in regression
                tests to hide the user-dependent portion of paths, so that logs would otherwise display a path
                like:

                'C:/Users/aleja/Documents/Code/chateauclaudia-labs/apodeixi/test-knowledge-base/envs/big_rocks_posting_ENV/excel-postings'

                instead display a "masked" path where the user-specific prefix is masked, so that only the
                logical portion of the path (logical as in: it is the structure mandated by the KnowledgeStore)
                is displayed. In the above example, that might become:

                '<MASKED>/envs/big_rocks_posting_ENV/excel-postings'
    '''
    def __init__(self, parent_trace, path_mask):
        self.functional_purpose     = None
        self.parent_trace           = parent_trace # Caller's FunctionalTrace

        self.path_mask              = path_mask

    ACTIVITY                        = 'activity'
    DATA                            = 'data'
    FLOW_STAGE                      = 'flow_stage'
    ORIGINATION                     = 'origination'

    def doing(self, activity, flow_stage=None, data=None, origination=None):
        '''
        Meant for a caller to create a new FunctionalTrace object that it can pass to a subroutine it is calling, 
        recording the step (`activity`) the caller is in the midst of processing as it invokes the subroutine.
        .
        The caller can optionally indicate some information about its flow to situate such activity. For example, it
        may set `flow_stage` to a message like "After validations of order A-123 are complete".

        Anything else worth recording can be put in the dictionary `data`

        @param activity         A string that identifies the activity the caller is in the midst of processing
        @param flow_stage       An optional string for caller to say where in its flow this activity sits (e.g., "after such and such" or
                                "before such and such")
        @param data             A dictionary or useful information about how the activity is being run (e.g., "4th cycle of loop" or
                                "in cycle of loop processing item ABC")
        @param 'origination': a dictionary of stack-related information (may include folder information that varies per installation)
        '''
        if activity==None:
            raise ApodeixiError(self.parent_trace, "Can't create a FunctionalTrace with a null activity")
        if type(activity) != str:
            raise ApodeixiError(self.parent_trace, "Can't create a FunctionalTrace with a non-string activity of type " + str.type(activity))
        if len(activity.strip('\n '))==0:
            raise ApodeixiError(self.parent_trace, "Can't create a FunctionalTrace with a blank activity '" + activity + "'")        

        # Ensure context is non-null
        if data == None:
            data                            = {}
        if flow_stage == None:
            flow_stage                      = ''
        if origination == None:
            origination                     = {}

        subroutine_ctx                      = FunctionalTrace(parent_trace=self, path_mask=self.path_mask)
        subroutine_ctx.functional_purpose   = { FunctionalTrace.ACTIVITY        : activity, 
                                                FunctionalTrace.FLOW_STAGE      : flow_stage,
                                                FunctionalTrace.DATA            : data,
                                                FunctionalTrace.ORIGINATION     : origination}

        return subroutine_ctx

    def examine(self, as_string=False, exclude_origination=False):
        '''
        Can be thought of as analogous to a (technical) stack trace, but expressed in functional terms.

        Specifically, this method produces a top-down list of runtime "functional intentions and context" that lead to the point in the
        code when this is invoked. The index nth of the list can be considered "Level n" for nexted functional contexts, where
        "Level 0" is the most general, root level.

        The "functional intentions and context" (i.e., the members of the list returned) are dictionaries with three keys:
        * 'activity:   a string describing the functional intention at the current level
        * 'flow_stage': a string describing where in the current level's functional flow we are
        * 'data':       a dictionary of potentially other useful information.
        * 'origination': a dictionary of stack-related information (may include folder information that varies per installation)

        Dependingon the `as_string` flag, this method either returns such list of dictionaries or a string that formats
        them nicely for display on a termina.
        '''
        trace                   = []
        if as_string:
            trace.append(self._format_functional_trace(exclude_origination))
        else:
            trace.append(self.functional_purpose)

        if self.parent_trace != None and self.parent_trace.functional_purpose != None:
            trace               = self.parent_trace.examine(as_string=as_string, exclude_origination=exclude_origination) + trace
        
        return trace

    def _format_functional_trace(self, exclude_origination):
        '''
        Returns a human-readable string with the content of the functional trace
        '''
        def MSK(txt): # Abbreviaation for the masking logic
            if self.path_mask != None:
                return self.path_mask(txt)
            else:
                return txt

        if self.functional_purpose == None:
            return ''
        result                  = ''
        result                  += '---->\tactivity\t'  + MSK(self.functional_purpose[FunctionalTrace.ACTIVITY]) + '\n'
        
        flow_stage              = self.functional_purpose[FunctionalTrace.FLOW_STAGE]
        if flow_stage != None and len(flow_stage.strip()) > 0:
            result              += '\n' + MSK(FunctionalTrace._ins(FunctionalTrace.FLOW_STAGE)) + ': ' \
                                    + MSK(self.functional_purpose[FunctionalTrace.FLOW_STAGE])
        data                    = self.functional_purpose[FunctionalTrace.DATA]
        for k in data.keys():
            result              += '\n' + MSK(FunctionalTrace._ins(k)) + ': ' + MSK(str(data[k]))

        origination             = self.functional_purpose[FunctionalTrace.ORIGINATION]
        if not exclude_origination:
            for k in origination.keys():
                result      += '\n' + MSK(FunctionalTrace._ins(k)) + ': ' + MSK(str(origination[k]))
        return result

    def _ins(txt):
        '''
        Helper method that returns a string of 32 characters, starting with 12 spaces and then txt followed by padding
        '''
        allowed_txt_length  = min (20, len(txt))
        allowed_txt         = txt[0:allowed_txt_length]
        padding_length      = 20 - allowed_txt_length
        padding             = ' ' * padding_length
        indentation         = ' ' * 12

        return indentation + allowed_txt + padding

class ApodeixiError (Exception):
    '''
    Error class recommended in Apodeixi. It extends ValueError with a FunctionalTrace object to make troubleshooting
    easier.

    @param external_stacktrace A string, containing the stacktrace for a component external to Apodeixi which
        resulted in failure. The component is "external" in the logical sense - it may well be in-process with
        Apodeixi. What makes it "external" is that it is code to which Apodeixi cedes control as a "black box",
        and if the "black box" fails Apodeixi raises a new ApodeixiError from the point at which the "black box"
        was invoked by Apodeixi.
        Thus, the ApodeixiError's stack trace only can tell us when Apodeixi knew something went wrong, but not
        the root cause of the problem within the external code. To assist debugging such problems, the construction
        of an ApodeixiError allows for attaching the external code's stack trace as an optional string. 
    '''
    def __init__(self, functional_trace, msg, data={}, origination={}, external_stacktrace=None):
        super().__init__(self, msg)
        self.functional_trace           = functional_trace
        self.msg                        = msg
        self.data                       = data
        self.origination                = origination
        self.external_stacktrace        = external_stacktrace

    def trace_message(self, exclude_stack_trace=False):

        def MSK(txt): # Abbreviation for the masking logic
            if self.functional_trace.path_mask != None:
                return self.functional_trace.path_mask(txt)
            else:
                return txt

        data_msg                    = ''
        for k in self.data.keys():
            data_msg                += '\n' + MSK(FunctionalTrace._ins(k)) + ': ' + MSK(str(self.data[k]))
        if not exclude_stack_trace:
            for k in self.origination.keys():
                data_msg            += '\n' + MSK(FunctionalTrace._ins(k)) + ': ' + MSK(str(self.origination[k]))
        if len(data_msg) > 0:
            data_msg                = '\n' + data_msg + '\n'

        advertisement_for_stack_trace = ''
        if not exclude_stack_trace:
            if self.external_stacktrace == None:
                advertisement_for_stack_trace = ' (stack trace at the bottom)'
            else:
                advertisement_for_stack_trace = '\n(failure occurred in an external component. So 2 strack traces are '\
                                    + 'provided at the bottom: '\
                                    + '\n\t-one originating from Apodeixi\'s invocation of the external component, '\
                                    + '\n\t-and one for the external component itself)'
        trace_msg               = '\n\n******** Functional Trace ********\n\n' + 'Problem:\t' + self.msg + data_msg \
                                    + '\nHere are the functional activities that led to the problem' \
                                    +  advertisement_for_stack_trace + ':' \
                                    + '\n\n' \
                                    + '\n\n'.join([str(trace_level) for trace_level in self.functional_trace.examine(as_string=True,
                                                                                        exclude_origination=exclude_stack_trace)]) \
                                    + '\n'   
        if not exclude_stack_trace:
            traceback_stream        = StringIO()
            _traceback.print_exc(file = traceback_stream)
            if self.external_stacktrace == None:
                trace_msg           += "\n" + "-"*60 + '\tTechnical Stack Trace\n\n'            
                trace_msg           += traceback_stream.getvalue()
                trace_msg           += "\n" + "-"*60  
            else:
                trace_msg           += "\n" + "-"*60 + '\tTechnical Stack Trace 1 of 2 (in external component)\n\n' 
                trace_msg           += self.external_stacktrace
                trace_msg           += "\n" + "-"*60 + '\tTechnical Stack Trace 2 of 2 (within Apodeixi)\n\n' 
                trace_msg           += traceback_stream.getvalue()
                trace_msg           += "\n" + "-"*60  
            
        return trace_msg

    