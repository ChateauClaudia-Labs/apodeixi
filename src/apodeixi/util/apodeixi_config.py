import os                                           as _os
import toml                                         as _toml

from apodeixi.util.a6i_error                        import ApodeixiError

from apodeixi.util.dictionary_utils     import DictionaryUtils # Temporary - will refactor to upstream

class ApodeixiConfig():
    '''
    Data structure to load and hold in memory the configuration settings for the apodeixi runtime
    '''
    def __init__(self, parent_trace):
        
        CONFIG_FILE                 = self._get_config_filename(parent_trace)
        CONFIG_FOLDER               = self._get_config_folder(parent_trace)

        my_trace                    = parent_trace.doing("Attempting to load Apodeixi configuration",
                                                            data = {'CONFIG_FOLDER': CONFIG_FOLDER, 'CONFIG_FILE': CONFIG_FILE})
        try:
            with open(CONFIG_FOLDER + '/' + CONFIG_FILE, 'r') as file:
                config_txt            = str(file.read())
            config_dict = _toml.loads(config_txt)
        except Exception as ex:
            raise ApodeixiError(my_trace, "Was not able to retrieve Apodeixi configuration due to: " + str(ex))

        self.config_dict            = config_dict

        my_trace                    = parent_trace.doing("Checking for any includes")
        check, explanation = DictionaryUtils().validate_path(   parent_trace    = my_trace, 
                                                                root_dict       = self.config_dict, 
                                                                root_dict_name  = 'apodeixi_config',
                                                                path_list       = ['include', 'file'],
                                                                valid_types     = [str])
        if check:
            INCLUDE_FILE            = config_dict['include']['file']
            try:
                my_trace                    = parent_trace.doing("Attemping to load include file referenced in Apodeixi configuration",
                                                                    data = {'include.file': INCLUDE_FILE})
                with open(CONFIG_FOLDER + '/' + INCLUDE_FILE, 'r') as file:
                    file_to_include_txt            = str(file.read())
                file_to_include_dict = _toml.loads(file_to_include_txt)
            except Exception as ex:
                raise ApodeixiError(my_trace, "Was not able to retrieve include file referenced inApodeixi configuration due to: " 
                                                + str(ex),
                                                data = {'include.file': INCLUDE_FILE})

            self.config_dict        = file_to_include_dict | self.config_dict

        # Determines wheter referential integrity checks are enforced. Should be True in production but
        # unit tests may choose to turn it off.
        self.enforce_referential_integrity  = True

        # This flag will be set by test cases to assist with masking non-deterministic information about the
        # location of the test database. It is used in the masking function that hides parts of paths from regression
        # output, to avoid non-deterministic test output. When not using the test regression suite, this flag plays no role.
        self.test_db_dir                    = None

    def _get_config_folder(self, parent_trace):
        APODEIXI_CONFIG_DIRECTORY                           = _os.environ.get('APODEIXI_CONFIG_DIRECTORY')

        if APODEIXI_CONFIG_DIRECTORY == None or len(APODEIXI_CONFIG_DIRECTORY.strip())==0:
            raise ApodeixiError(parent_trace, "Environment variable $APODEIXI_CONFIG_DIRECTORY is not set, so can't " 
                                                + "figure out from which folder to load the Apodeixi configuration")

        if not _os.path.isdir(APODEIXI_CONFIG_DIRECTORY):
            raise ApodeixiError(parent_trace, "Environment variable $APODEIXI_CONFIG_DIRECTORY does not point to a valid directory: '"
                                                + str(APODEIXI_CONFIG_DIRECTORY) + "'")

        return APODEIXI_CONFIG_DIRECTORY

    def _get_config_filename(self, parent_trace):
        '''
        Should be extended by derived classes to change where the configuration file should come from
        '''
        return 'apodeixi_config.toml'

    def get_static_data_api(self, parent_trace):
        '''
        Returns the Manifest API for static data. Typical usage is for code that needs to retrieve static
        data manifests.
        '''
        # This is hard-coded on purpose. Other systems that extend Apodeixi may have different APIs for static
        # data, so by making this a function that is overwritable by derived non-Apodeixi classes, we ensure
        # that Apodeixi code like the StaticDataValidator still works, since it can query this method to find
        # out what manifest API corresponds to static data.
        STATIC_DATA_API     = 'static-data.admin.a6i.io'
        return STATIC_DATA_API

    def getSecretsFolder(self, parent_trace):
        my_trace            = parent_trace.doing("Retrieving the secrets' folder from the Apodeixi Configuration ")
        check, explanation = DictionaryUtils().validate_path(   parent_trace    = my_trace, 
                                                                root_dict       = self.config_dict, 
                                                                root_dict_name  = 'apodeixi_config',
                                                                path_list       = ['secrets', 'folder'],
                                                                valid_types     = [str])
        if not check:
            raise ApodeixiError(my_trace, "Can't locate secrets' folder: " + explanation)
        
        # Expand any environment variables in the path
        return _os.path.expandvars(self.config_dict['secrets']['folder'])    

    def get_KB_RootFolder(self, parent_trace): 
        my_trace            = parent_trace.doing("Retrieving Knowledge Base's root folder from the Apodeixi Configuration ")
        KB                  = 'knowledge-base'
        ROOT_FOLDER         = 'knowledge-base-root-folder'
        check, explanation = DictionaryUtils().validate_path(   parent_trace    = my_trace, 
                                                                root_dict       = self.config_dict, 
                                                                root_dict_name  = 'apodeixi',
                                                                path_list       = [KB, ROOT_FOLDER],
                                                                valid_types     = [str])
        if not check:
            raise ApodeixiError(my_trace, "Can't locate root folder for Knowledge Base: " + explanation)
        
        # Expand any environment variables in the path
        return _os.path.expandvars(self.config_dict[KB][ROOT_FOLDER])

    def get_ExternalCollaborationFolder(self, parent_trace):
        my_trace            = parent_trace.doing("Retrieving external collaboration root folder from the "
                                                    + "Apodeixi Configuration ")
        KB                  = 'knowledge-base'
        EXTERNAL_FOLDER     = 'external-collaboration-folder'
        check, explanation = DictionaryUtils().validate_path(   parent_trace    = my_trace, 
                                                                root_dict       = self.config_dict, 
                                                                root_dict_name  = 'apodeixi',
                                                                path_list       = [KB, EXTERNAL_FOLDER],
                                                                valid_types     = [str])
        if not check:
            raise ApodeixiError(my_trace, "Can't locate external collaboration folder: " + explanation)
        
        # Expand any environment variables in the path
        return _os.path.expandvars(self.config_dict[KB][EXTERNAL_FOLDER])

    def getMonthFiscalYearStarts(self, parent_trace):
        my_trace            = parent_trace.doing("Retrieving Knowledge Base's fiscal year start from the Apodeixi Configuration ")
        SETTINGS            = 'organization-settings'
        FY_START            = 'month-fiscal-year-starts'
        check, explanation = DictionaryUtils().validate_path(   parent_trace    = my_trace, 
                                                                root_dict       = self.config_dict, 
                                                                root_dict_name  = 'apodeixi',
                                                                path_list       = [SETTINGS, FY_START],
                                                                valid_types     = [int])
        if not check:
            raise ApodeixiError(my_trace, "Can't locate Knowledge Base's fiscal year start: " + explanation)
        
        return self.config_dict[SETTINGS][FY_START]

    def getOrganization(self, parent_trace):
        my_trace            = parent_trace.doing("Retrieving Knowledge Base's organization from the Apodeixi Configuration ")
        SETTINGS            = 'organization-settings'
        ORGANIZATION        = 'organization'
        check, explanation = DictionaryUtils().validate_path(   parent_trace    = my_trace, 
                                                                root_dict       = self.config_dict, 
                                                                root_dict_name  = 'apodeixi',
                                                                path_list       = [SETTINGS, ORGANIZATION],
                                                                valid_types     = [str])
        if not check:
            raise ApodeixiError(my_trace, "Can't locate Knowledge Base's organization: " + explanation)
        
        return self.config_dict[SETTINGS][ORGANIZATION]

    def getKnowledgeBaseAreas(self, parent_trace):
        my_trace            = parent_trace.doing("Retrieving Knowledge Base's areas from the Apodeixi Configuration ")
        SETTINGS            = 'organization-settings'
        ORGANIZATION        = 'knowledge-base-areas'
        check, explanation = DictionaryUtils().validate_path(   parent_trace    = my_trace, 
                                                                root_dict       = self.config_dict, 
                                                                root_dict_name  = 'apodeixi',
                                                                path_list       = [SETTINGS, ORGANIZATION],
                                                                valid_types     = [list])
        if not check:
            raise ApodeixiError(my_trace, "Can't locate Knowledge Base's areas: " + explanation)
        
        return self.config_dict[SETTINGS][ORGANIZATION]

    def get_CLI_InitializerClassname(self, parent_trace):
        my_trace            = parent_trace.doing("Retrieving Knowledge Base's CLI initializer classname from the Apodeixi Configuration ")
        CLI                 = 'cli'
        CLASSNAME           = 'initializer-classname'
        check, explanation = DictionaryUtils().validate_path(   parent_trace    = my_trace, 
                                                                root_dict       = self.config_dict, 
                                                                root_dict_name  = 'apodeixi',
                                                                path_list       = [CLI, CLASSNAME],
                                                                valid_types     = [str])
        if not check:
            raise ApodeixiError(my_trace, "Can't locate Knowledge Base's areas: " + explanation)
        
        return self.config_dict[CLI][CLASSNAME]

    def getGrandfatheredScoringCycles(self, parent_trace):
        '''
        Returns a list of strings, corresponding to manually configured scoring cycles that are considered
        valid.
        This method exists for backward compatibility reasons, to not invalidate data created before Apodeixi started
        enforcing that scoring cycles must be strings that can be successfully parsed into FY_Quarter objects.
        '''
        my_trace            = parent_trace.doing("Retrieving grandfathered scorcing cycles from the Apodeixi Configuration ")
        BACK_COMPATIBILITY  = 'backward-compabitility'
        GRANDFATHERED_SC    = 'grandfathered_scoring_cycles'
        check, explanation = DictionaryUtils().validate_path(   parent_trace    = my_trace, 
                                                                root_dict       = self.config_dict, 
                                                                root_dict_name  = 'apodeixi',
                                                                path_list       = [BACK_COMPATIBILITY, GRANDFATHERED_SC],
                                                                valid_types     = [list])
        if not check:
            # If nothing is grandfathered, that is good. It means this deployment will only use modern
            # scoring cycles that can be converted to FY_Quarter objects
            return []
        
        return self.config_dict[BACK_COMPATIBILITY][GRANDFATHERED_SC]