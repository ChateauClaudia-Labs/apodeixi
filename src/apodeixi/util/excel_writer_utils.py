from apodeixi.util.formatting_utils     import DictionaryFormatter

def nice_ws_info(parent_trace, worksheet_info):
    '''
    Helper method to create a "nice string" that can be outputted in a readable form (for example, in test outputs)

    param worksheet_info An instance of apodeixi.representers.as_excel.XL_WorksheetInfo that should be formatted as a string
    '''
    nice_format                     = ''
    nice_format += "\n======================== Column information =========================="
    nice_format += DictionaryFormatter().dict_2_nice(   parent_trace    = parent_trace,
                                                        a_dict          = worksheet_info.colinfo)

    fmt_dict                        = worksheet_info.format_dict
    for row_nb in fmt_dict.keys():
        row_dict                    = fmt_dict[row_nb]
        for col_nb in row_dict.keys():
            nice_format += "\n\n================ Formats row = " + str(row_nb) + ", col = " + str(col_nb) + " ============"
            cell_fmt_dict           = row_dict[col_nb]
            nice                    = DictionaryFormatter().dict_2_nice(    parent_trace    = parent_trace,
                                                                            a_dict          = cell_fmt_dict)
            nice_format += "\n" + nice
    return nice_format


