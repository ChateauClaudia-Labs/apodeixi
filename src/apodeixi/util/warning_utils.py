import sys                                              as _sys
import pandas                                           as _pd
import warnings
import traceback
from io                                                 import StringIO

from apodeixi.util.a6i_error                            import ApodeixiError

class ApodeixiWarningMessage(warnings.WarningMessage):
    '''
    Helper data structure used by WarningUtils. It extends the parent class with additional stack trace
    information that Apodeixi is interested in getting for warnings that get raised by dependencies used by
    Apodeixi, to assist the developer in understanding where the warnings are being raised from

    @param stacktrace: A string, representing the stack trace from which a warning was raised
    '''
    def __init__(self, message, stacktrace, category, filename, lineno, file=None, line=None, source=None):
        super().__init__(message, category, filename, lineno, file, line, source)

        self.stacktrace             = stacktrace

class WarningUtils():

    def __init__(self):
        pass

    def turn_traceback_on(self, parent_trace, warnings_list):
        '''
        This method ensures that warnings print stack traces.

        Returns a StringIO object, to which stack traces are printed, that the caller can then examine, print, or dispose
        of as appropriate.

        The intention is to pin down the origin of warnings, in the case of warnings that should be caught in Apodeixi.
        For example, noisy warning that must be ignored, or serious warnings that should become errors.

        One of the motivations is to prevent warnings from creeping into test regression output, particularly in the CLI, 
        where that noise may be non-deterministic and cause spurious test failures.

        Implementation is as suggested in https://stackoverflow.com/questions/22373927/get-traceback-of-warnings 

        @param warnings_list A list, as created by the warnings.catch_warnings context manager, to which we will be
            appending each warning's message as an element to that list.
        '''
        def _warn_with_traceback(message, category, filename, lineno, file=None, line=None):

            log                     = StringIO()
            traceback.print_stack(file=log)
            log.write(warnings.formatwarning(message, category, filename, lineno, line))
            stacktrace              = log.getvalue()
            warn_msg                = ApodeixiWarningMessage(message, stacktrace, category, filename, lineno, file, line)
            warnings_list.append(warn_msg)

        warnings.showwarning        = _warn_with_traceback 

    def check_if_user_error(self, parent_trace, a_warning):
        '''
        Helper method for this class to handle certain warnings that Apodeixi considers user errors, as
        opposed to development-type errors.

        Normally 3rd-party warnings should be pre-empted at development time by the Apodeixi developers
        using those 3rd party dependencies correctly, avoiding code constructs that the 3rd party flags as
        risky or deprecated.

        However, in some cases the warning is really a user error, so the remedy lies with the user. However,
        the warning is probably very unfriendly to the user, so it should be converted to something more
        Apodeixi-like and less about the 3rd party dependency.

        That is the purpose of this method: if a warning is user error, it will raise an ApodeixiError with
        as friendly a message as possible. Otherwise, it does nothing.
        '''
        if str(a_warning.message).startswith("Defining usecols with out of bounds indices is deprecated"):
            # This is Pandas warning that happens when the user submits an Excel posting with a range of
            # columns that includes blank columns. For example, the user might submit H2:L30 as a range,
            # but if column L is blank, then the user should instead have submitted H2:K30
            #
            warning_dict        = {"Problem encountered": a_warning.message}
            raise ApodeixiError(parent_trace, "Something went wrong, most probably because you might be attempting "
                                            + "to post an Excel file and one of the PostingLabel's ranges includes empty "
                                            + "Excel columns. If so, please correct your PostingLabel and try again.",
                                data = warning_dict,
                                external_stacktrace     = a_warning.stacktrace)

        else:
            pass

    def check_if_should_ignore(self, parent_trace, a_warning):
        '''
        Helper method for this class to handle certain warnings from 3rd party libraries that Apodeixi should ignore.
        
        This method returns True if the warning should be ignored, and False otherwise
        '''
        if a_warning.category == ResourceWarning and str(a_warning.message).startswith("unclosed event loop"):
            #Ignore such warnings - they are noise generated by YAML
            return True

        elif a_warning.category == UserWarning and str(a_warning.message).startswith("Data Validation extension is not supported and will be removed"):
            # Ignore such warnings - they are noise generated by openpyxl that Pandas just transmits up to
            # Apodeixi. So it is "Pandas noise", i.e., Pandas is using openpyxl in a way that triggers warnings,
            # deep down and hidden from the Pandas API that Apodeixi gets to see. So not much Apodeixi can do other
            # than ignore it.
            return True
        elif a_warning.category == _pd.errors.PerformanceWarning and str(a_warning.message).startswith("dropping on a non-lexsorted multi-index without a level parameter may impact performance"):
            # Pandas generates this when we group by a column that is not a tuple when other columns are tuples
            # So ignore it - our reporting routinely does such group-by operations
            return True
        else:
            return False

    def handle_warnings(self, parent_trace, warning_list):
        '''
        Helper method to catch warnings and turns them into ApodeixiErrors.
        '''   
        if len(warning_list) > 0:
            warning_dict                                = {}
            for idx in range(len(warning_list)):
                a_warning                               = warning_list[idx]
                self.check_if_user_error(parent_trace, a_warning)
                if self.check_if_should_ignore(parent_trace, a_warning):
                    # In this case, this particular warning should not trigger an ApodeixiError.
                    # So ignore it and move on to examining the next warning
                    continue

                warning_dict['Warning ' + str(idx)]     = str(a_warning.message)
                warning_dict['... from']                = str(a_warning.filename)
                warning_dict['... at line']             = str(a_warning.lineno)

                trace_msg                               = "\n" + "-"*60 + "\tWarnings Stack Trace\n\n"
                trace_msg                               += str(a_warning.stacktrace)
                trace_msg                               += "\n" + "-"*60
                warning_dict['Stack Trace']             = trace_msg

            if len(warning_dict.keys()) > 0:
                raise ApodeixiError(parent_trace, "A dependency issued " + str(len(warning_list)) + " warning(s)",
                                    data = {} | warning_dict)  
