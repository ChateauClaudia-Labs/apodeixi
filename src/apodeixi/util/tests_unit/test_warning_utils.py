import sys                                          as _sys
import warnings

from apodeixi.testing_framework.a6i_unit_test       import ApodeixiUnitTest
from apodeixi.util.a6i_error                        import ApodeixiError, FunctionalTrace

from apodeixi.util.warning_utils                    import WarningUtils

class Test_WarningUtils(ApodeixiUnitTest):

    def setUp(self):
        super().setUp()

    def test_warning_utils(self):
        root_trace                          = FunctionalTrace(parent_trace=None, path_mask=self._path_mask).doing("Testing Warning Utils")
        try:
            TEST_SCENARIO                   = 'test_warning_utils'

            my_trace                        = root_trace.doing("Testing a fake warning")

            with warnings.catch_warnings(record=True) as w:
                WarningUtils().turn_traceback_on(my_trace, warnings_list=w)
                    
                warnings.warn("Test warning for Warning Utils", DeprecationWarning)

                WarningUtils().handle_warnings(my_trace, warning_list=w)

                # The handling of the warning should raise an exception, so we should not get here
                self.assertTrue(1==2)

        except ApodeixiError as ex:
            output_txt                      = ex.trace_message()
            self._compare_to_expected_txt(  parent_trace        = my_trace,
                                            output_txt          = output_txt,
                                            test_output_name    = TEST_SCENARIO, 
                                            save_output_txt     = True)

if __name__ == "__main__":
    # execute only if run as a script
    def main(args):
        T = Test_WarningUtils()
        T.setUp()
        what_to_do = args[1]
        if what_to_do=="warning_utils":
            T.test_warning_utils()
        T.tearDown()
        
    main(_sys.argv)