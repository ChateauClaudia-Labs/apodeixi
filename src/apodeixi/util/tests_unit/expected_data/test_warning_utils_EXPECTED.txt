

******** Functional Trace ********

Problem:	A dependency issued 1 warning(s)

            Warning 0           : Test warning for Warning Utils
            ... from            : <APODEIXI INSTALLATION>/apodeixi\util\tests_unit\test_warning_utils.py
            ... at line         : 24
            Stack Trace         : 
------------------------------------------------------------	Warnings Stack Trace

  File "C:\Users\aleja\Documents\CodeImages\Technos\Anaconda3\envs\a6i-env\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\aleja\Documents\CodeImages\Technos\Anaconda3\envs\a6i-env\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\aleja\Documents\CodeImages\Technos\Anaconda3\envs\a6i-env\lib\unittest\__main__.py", line 18, in <module>
    main(module=None)
  File "C:\Users\aleja\Documents\CodeImages\Technos\Anaconda3\envs\a6i-env\lib\unittest\main.py", line 101, in __init__
    self.runTests()
  File "C:\Users\aleja\Documents\CodeImages\Technos\Anaconda3\envs\a6i-env\lib\unittest\main.py", line 271, in runTests
    self.result = testRunner.run(self.test)
  File "C:\Users\aleja\Documents\CodeImages\Technos\Anaconda3\envs\a6i-env\lib\unittest\runner.py", line 176, in run
    test(result)
  File "C:\Users\aleja\Documents\CodeImages\Technos\Anaconda3\envs\a6i-env\lib\unittest\suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "C:\Users\aleja\Documents\CodeImages\Technos\Anaconda3\envs\a6i-env\lib\unittest\suite.py", line 122, in run
    test(result)
  File "C:\Users\aleja\Documents\CodeImages\Technos\Anaconda3\envs\a6i-env\lib\unittest\suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "C:\Users\aleja\Documents\CodeImages\Technos\Anaconda3\envs\a6i-env\lib\unittest\suite.py", line 122, in run
    test(result)
  File "C:\Users\aleja\Documents\CodeImages\Technos\Anaconda3\envs\a6i-env\lib\unittest\suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "C:\Users\aleja\Documents\CodeImages\Technos\Anaconda3\envs\a6i-env\lib\unittest\suite.py", line 122, in run
    test(result)
  File "C:\Users\aleja\Documents\CodeImages\Technos\Anaconda3\envs\a6i-env\lib\unittest\case.py", line 651, in __call__
    return self.run(*args, **kwds)
  File "C:\Users\aleja\Documents\CodeImages\Technos\Anaconda3\envs\a6i-env\lib\unittest\case.py", line 592, in run
    self._callTestMethod(testMethod)
  File "C:\Users\aleja\Documents\CodeImages\Technos\Anaconda3\envs\a6i-env\lib\unittest\case.py", line 550, in _callTestMethod
    method()
<APODEIXI INSTALLATION>/apodeixi\util\tests_unit\test_warning_utils.py", line 24, in test_warning_utils
    warnings.warn("Test warning for Warning Utils", DeprecationWarning)
  File "C:\Users\aleja\Documents\CodeImages\Technos\Anaconda3\envs\a6i-env\lib\warnings.py", line 109, in _showwarnmsg
    sw(msg.message, msg.category, msg.filename, msg.lineno,
<APODEIXI INSTALLATION>/apodeixi\util\warning_utils.py", line 48, in _warn_with_traceback
    traceback.print_stack(file=log)
<APODEIXI INSTALLATION>/apodeixi\util\tests_unit\test_warning_utils.py:24: DeprecationWarning: Test warning for Warning Utils
  warnings.warn("Test warning for Warning Utils", DeprecationWarning)

------------------------------------------------------------

Here are the functional activities that led to the problem (stack trace at the bottom):

---->	activity	Testing Warning Utils


---->	activity	Testing a fake warning


------------------------------------------------------------	Technical Stack Trace

Traceback (most recent call last):
  File "C:\Users\aleja\Documents\Code\chateauclaudia-labs\apodeixi\project\src\apodeixi\util\tests_unit\test_warning_utils.py", line 26, in test_warning_utils
    WarningUtils().handle_warnings(my_trace, warning_list=w)
  File "C:\Users\aleja\Documents\Code\chateauclaudia-labs\apodeixi\project\src\apodeixi\util\warning_utils.py", line 134, in handle_warnings
    raise ApodeixiError(parent_trace, "A dependency issued " + str(len(warning_list)) + " warning(s)",
apodeixi.util.a6i_error.ApodeixiError: (ApodeixiError(...), 'A dependency issued 1 warning(s)')

------------------------------------------------------------